File: mac/DeskLayer/DeskLayer/Core/HookServer.swift
//
//  HookServer.swift
//  DeskLayer
//
//  One app-level, loopback-ONLY HTTP listener. Local tools (Claude/Codex
//  hooks, scripts) POST to it; the server fans each request out to every
//  running plugin that registered a matching handler via $server.on(...).
//  Never bound to an external interface.
//
//  Owned by RuntimeCoordinator; plugins never open their own port. The port
//  is configurable (DESKLAYER_HOOK_PORT env var, or the DeskLayer.hookPort
//  default: `defaults write com.qiudaomao.DeskLayer DeskLayer.hookPort 9787`)
//  and the listener only binds while at least one plugin has a handler
//  registered — no $server plugins, no open port to conflict with anyone.
//

import Foundation
import Network
import os

nonisolated final class HookServer: @unchecked Sendable {
    struct Handler {
        let itemID: UUID
        let method: String
        /// Delivers (event, body) on the plugin's own queue.
        let deliver: @Sendable ([String: Any], String) -> Void
    }

    /// The loopback port to bind when a handler exists:
    /// DESKLAYER_HOOK_PORT env var > DeskLayer.hookPort default > 8787.
    static func resolvedPort() -> UInt16 {
        if let text = ProcessInfo.processInfo.environment["DESKLAYER_HOOK_PORT"],
           let value = UInt16(text), value > 0 {
            return value
        }
        let stored = UserDefaults.standard.integer(forKey: "DeskLayer.hookPort")
        if stored > 0, stored <= Int(UInt16.max) { return UInt16(stored) }
        return 8787
    }

    private let configuredPort: UInt16
    private let lock = NSLock()
    private var handlers: [Handler] = []
    private var listener: NWListener?
    private(set) var port: UInt16?
    private let log = Logger(subsystem: "com.qiudaomao.DeskLayer", category: "hookserver")

    init(port: UInt16 = HookServer.resolvedPort()) {
        configuredPort = port
    }

    // MARK: - Lifecycle (driven by handler registration; callers never start)

    /// Bind to the configured loopback port. Caller holds `lock`.
    private func startLocked() {
        guard listener == nil else { return }
        guard let nwPort = NWEndpoint.Port(rawValue: configuredPort) else { return }
        let parameters = NWParameters.tcp
        parameters.requiredLocalEndpoint = NWEndpoint.hostPort(host: .ipv4(.loopback), port: nwPort)
        guard let newListener = try? NWListener(using: parameters) else {
            log.error("failed to bind 127.0.0.1:\(self.configuredPort)")
            return
        }
        newListener.newConnectionHandler = { [weak self] connection in
            self?.accept(connection)
        }
        newListener.stateUpdateHandler = { [weak self] state in
            if case .failed(let error) = state {
                self?.log.error("listener failed: \(error.localizedDescription, privacy: .public)")
            }
        }
        newListener.start(queue: DispatchQueue.global(qos: .utility))
        listener = newListener
        port = configuredPort
        log.info("hook server listening on 127.0.0.1:\(self.configuredPort)")
    }

    /// Caller holds `lock`.
    private func stopLocked() {
        guard listener != nil else { return }
        listener?.cancel()
        listener = nil
        port = nil
        log.info("hook server stopped (no handlers)")
    }

    func stop() {
        lock.lock()
        stopLocked()
        lock.unlock()
    }

    // MARK: - Registration (called from plugin instances)

    /// The first handler brings the listener up; the last one down. A rebuild
    /// (removeAll then re-register) therefore releases the port only while no
    /// $server plugin is running.
    func addHandler(_ handler: Handler) {
        lock.lock()
        handlers.append(handler)
        startLocked()
        lock.unlock()
    }

    func removeHandlers(itemID: UUID) {
        lock.lock()
        handlers.removeAll { $0.itemID == itemID }
        if handlers.isEmpty { stopLocked() }
        lock.unlock()
    }

    /// Cleared synchronously at the start of a runtime rebuild, so a stale
    /// instance's async teardown can't drop a freshly re-registered handler.
    func removeAllHandlers() {
        lock.lock()
        handlers.removeAll()
        stopLocked()
        lock.unlock()
    }

    private func matching(method: String) -> [Handler] {
        lock.lock()
        defer { lock.unlock() }
        return handlers.filter { $0.method == method }
    }

    // MARK: - Connection handling

    private func accept(_ connection: NWConnection) {
        connection.start(queue: DispatchQueue.global(qos: .utility))
        receive(connection, buffer: Data())
    }

    private func receive(_ connection: NWConnection, buffer: Data) {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 64 * 1024) { [weak self] data, _, isComplete, error in
            guard let self, error == nil else { connection.cancel(); return }
            var buffer = buffer
            if let data { buffer.append(data) }
            if buffer.count > 1 << 20 { // 1 MB cap
                self.respond(connection, status: "413 Payload Too Large", json: ["ok": false, "error": "too large"])
                return
            }
            if let request = HTTPRequest(data: buffer) {
                self.dispatch(request, connection: connection)
            } else if isComplete {
                connection.cancel()
            } else {
                self.receive(connection, buffer: buffer)
            }
        }
    }

    private func dispatch(_ request: HTTPRequest, connection: NWConnection) {
        let targets = matching(method: request.method)
        let event: [String: Any] = [
            "method": request.method,
            "path": request.path,
            "headers": request.headers,
        ]
        for handler in targets {
            handler.deliver(event, request.body)
        }
        respond(connection, status: "200 OK", json: ["ok": true, "delivered": targets.count])
    }

    private func respond(_ connection: NWConnection, status: String, json: [String: Any]) {
        let body = (try? JSONSerialization.data(withJSONObject: json)) ?? Data("{}".utf8)
        let head = "HTTP/1.1 \(status)\r\nContent-Type: application/json\r\nContent-Length: \(body.count)\r\nConnection: close\r\n\r\n"
        connection.send(content: Data(head.utf8) + body, completion: .contentProcessed { _ in
            connection.cancel()
        })
    }
}

// MARK: - Minimal HTTP/1.1 request parsing

private struct HTTPRequest {
    let method: String
    let path: String
    let headers: [String: String]
    let body: String

    /// Returns nil while the request is still incomplete.
    init?(data: Data) {
        guard let headerEnd = data.range(of: Data("\r\n\r\n".utf8)) else { return nil }
        let headerText = String(decoding: data[..<headerEnd.lowerBound], as: UTF8.self)
        var lines = headerText.components(separatedBy: "\r\n")
        guard !lines.isEmpty else { return nil }
        let requestParts = lines.removeFirst().components(separatedBy: " ")
        guard requestParts.count >= 2 else { return nil }

        var parsedHeaders: [String: String] = [:]
        for line in lines {
            guard let colon = line.firstIndex(of: ":") else { continue }
            let key = line[..<colon].trimmingCharacters(in: .whitespaces).lowercased()
            let value = line[line.index(after: colon)...].trimmingCharacters(in: .whitespaces)
            parsedHeaders[key] = value
        }
        let contentLength = Int(parsedHeaders["content-length"] ?? "0") ?? 0
        let bodyData = data[headerEnd.upperBound...]
        guard bodyData.count >= contentLength else { return nil }

        method = requestParts[0].uppercased()
        path = requestParts[1]
        headers = parsedHeaders
        body = String(decoding: bodyData.prefix(contentLength), as: UTF8.self)
    }
}

File: win/src/DeskLayer.App/HookServer.cs
// One app-level, loopback-only HTTP listener — the Windows twin of the mac
// HookServer. Local tools POST to 127.0.0.1:<port>; each request fans out to
// every running plugin that registered a matching method via $server.on().
// A raw TcpListener (not HttpListener) avoids URL-ACL/admin prompts and
// never binds to an external interface.
//
// The port is configurable (DESKLAYER_HOOK_PORT env var, or "hookPort" in
// %APPDATA%\DeskLayer\settings.json; default 8787) and the listener only
// binds while at least one plugin has a handler registered — no $server
// plugins, no open port to conflict with anyone.

using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using DeskLayer.Core.Model;

namespace DeskLayer.App;

public sealed class HookServer : IDisposable
{
    public sealed record Handler(Guid ItemId, string Method, Action<string, string> Deliver);

    private readonly object gate = new();
    private readonly List<Handler> handlers = new();
    private readonly Action<string> log;
    private readonly int configuredPort;
    private TcpListener? listener;
    private CancellationTokenSource? cts;
    public int Port { get; private set; }

    public HookServer(Action<string> log, int? port = null)
    {
        this.log = log;
        configuredPort = port ?? ResolvePort();
    }

    /// The loopback port to bind when a handler exists:
    /// DESKLAYER_HOOK_PORT env var > settings.json "hookPort" > 8787.
    public static int ResolvePort()
    {
        if (Environment.GetEnvironmentVariable("DESKLAYER_HOOK_PORT") is { Length: > 0 } env &&
            int.TryParse(env, out var fromEnv) && fromEnv is > 0 and < 65536)
            return fromEnv;
        try
        {
            var path = Path.Combine(LayoutStore.DataDirectory, "settings.json");
            if (File.Exists(path) &&
                JsonNode.Parse(File.ReadAllText(path)) is JsonObject root &&
                root["hookPort"] is JsonValue value &&
                value.TryGetValue<int>(out var stored) && stored is > 0 and < 65536)
                return stored;
        }
        catch (Exception ex) when (ex is IOException or JsonException or UnauthorizedAccessException) { }
        return 8787;
    }

    // ---- lifecycle (driven by handler registration; callers never start) ----

    /// Bind to the configured loopback port. Caller holds `gate`.
    private void StartLocked()
    {
        if (listener != null) return;
        try
        {
            listener = new TcpListener(IPAddress.Loopback, configuredPort); // loopback only
            listener.Start();
            Port = configuredPort;
            cts = new CancellationTokenSource();
            _ = AcceptLoop(listener, cts.Token);
            log($"hook server listening on 127.0.0.1:{configuredPort}");
        }
        catch (SocketException ex)
        {
            log($"hook server failed to bind 127.0.0.1:{configuredPort}: {ex.Message}");
            listener = null;
            Port = 0;
        }
    }

    /// Caller holds `gate`.
    private void StopLocked()
    {
        if (listener == null) return;
        cts?.Cancel();
        try { listener.Stop(); } catch (SocketException) { }
        listener = null;
        cts = null;
        Port = 0;
        log("hook server stopped (no handlers)");
    }

    public void Stop()
    {
        lock (gate) StopLocked();
    }

    // ---- registration (called from plugin instances via the coordinator) ----

    /// The first handler brings the listener up; the last one down. A rebuild
    /// (RemoveAll then re-register) therefore releases the port only while no
    /// $server plugin is running.
    public void AddHandler(Handler handler)
    {
        lock (gate)
        {
            handlers.Add(handler);
            StartLocked();
        }
    }

    public void RemoveHandlers(Guid itemId)
    {
        lock (gate)
        {
            handlers.RemoveAll(h => h.ItemId == itemId);
            if (handlers.Count == 0) StopLocked();
        }
    }

    /// Cleared synchronously at the start of a runtime rebuild so a stale
    /// instance's teardown can't drop a freshly re-registered handler.
    public void RemoveAllHandlers()
    {
        lock (gate)
        {
            handlers.Clear();
            StopLocked();
        }
    }

    private List<Handler> Matching(string method)
    {
        lock (gate) return handlers.Where(h => h.Method == method).ToList();
    }

    // ---- connection handling ----

    private async Task AcceptLoop(TcpListener active, CancellationToken token)
    {
        while (!token.IsCancellationRequested)
        {
            TcpClient client;
            try { client = await active.AcceptTcpClientAsync(token); }
            catch (OperationCanceledException) { return; }
            catch (SocketException) { return; }
            _ = HandleConnection(client);
        }
    }

    private async Task HandleConnection(TcpClient client)
    {
        using (client)
        {
            try
            {
                using var stream = client.GetStream();
                var request = await ReadRequest(stream);
                if (request == null) return;

                var targets = Matching(request.Method);
                var eventJson = JsonSerializer.Serialize(new Dictionary<string, object>
                {
                    ["method"] = request.Method,
                    ["path"] = request.Path,
                    ["headers"] = request.Headers,
                });
                foreach (var handler in targets) handler.Deliver(eventJson, request.Body);

                var payload = JsonSerializer.Serialize(new { ok = true, delivered = targets.Count });
                var body = Encoding.UTF8.GetBytes(payload);
                var head = $"HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: {body.Length}\r\nConnection: close\r\n\r\n";
                await stream.WriteAsync(Encoding.UTF8.GetBytes(head));
                await stream.WriteAsync(body);
            }
            catch (Exception ex) when (ex is IOException or SocketException or ObjectDisposedException)
            {
                // Client hung up; nothing to do.
            }
        }
    }

    private sealed record HttpRequest(string Method, string Path, Dictionary<string, string> Headers, string Body);

    private static async Task<HttpRequest?> ReadRequest(NetworkStream stream)
    {
        var buffer = new byte[64 * 1024];
        var received = new MemoryStream();
        var headerEnd = -1;
        while (headerEnd < 0)
        {
            var read = await stream.ReadAsync(buffer);
            if (read == 0) return null;
            received.Write(buffer, 0, read);
            if (received.Length > 1 << 20) return null; // 1 MB cap
            headerEnd = IndexOfDoubleCrlf(received.GetBuffer(), (int)received.Length);
        }

        var all = received.GetBuffer();
        var headerText = Encoding.UTF8.GetString(all, 0, headerEnd);
        var lines = headerText.Split("\r\n");
        var requestParts = lines[0].Split(' ');
        if (requestParts.Length < 2) return null;

        var headers = new Dictionary<string, string>();
        foreach (var line in lines.Skip(1))
        {
            var colon = line.IndexOf(':');
            if (colon < 0) continue;
            headers[line[..colon].Trim().ToLowerInvariant()] = line[(colon + 1)..].Trim();
        }

        var contentLength = headers.TryGetValue("content-length", out var cl) && int.TryParse(cl, out var n) ? n : 0;
        var bodyStart = headerEnd + 4;
        while (received.Length - bodyStart < contentLength)
        {
            var read = await stream.ReadAsync(buffer);
            if (read == 0) break;
            received.Write(buffer, 0, read);
        }
        all = received.GetBuffer();
        var bodyLen = Math.Min(contentLength, (int)received.Length - bodyStart);
        var body = bodyLen > 0 ? Encoding.UTF8.GetString(all, bodyStart, bodyLen) : "";

        return new HttpRequest(requestParts[0].ToUpperInvariant(), requestParts[1], headers, body);
    }

    private static int IndexOfDoubleCrlf(byte[] data, int length)
    {
        for (var i = 0; i + 3 < length; i++)
            if (data[i] == '\r' && data[i + 1] == '\n' && data[i + 2] == '\r' && data[i + 3] == '\n')
                return i;
        return -1;
    }

    public void Dispose() => Stop();
}
